source/extensions/filters/network/common/redis/redis_command_stats.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/timespan.h"
+
+#include "common/stats/symbol_table_impl.h"
+
+#include "extensions/filters/network/common/redis/codec.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Common {
+namespace Redis {
+
+class RedisCommandStats {
+public:
+  RedisCommandStats(Stats::Scope& scope, const std::string& prefix, bool enabled);
+
+  Stats::Counter& counter(std::string name);
+  Stats::Histogram& histogram(std::string name);
+  Stats::Histogram& histogram(Stats::StatName stat_name);
+  Stats::CompletableTimespanPtr createCommandTimer(std::string name,
+                                                   Envoy::TimeSource& time_source);
+  Stats::CompletableTimespanPtr createAggregateTimer(Envoy::TimeSource& time_source);
+  std::string getCommandFromRequest(const RespValue& request);
+  void updateStatsTotal(std::string command);
+  void updateStats(const bool success, std::string command);
+  bool enabled() { return enabled_; }
+
+private:
+  void createStats(std::string name);
+  Stats::SymbolTable::StoragePtr addPrefix(const Stats::StatName name);
+
+  Stats::Scope& scope_;
+  Stats::StatNameSet stat_name_set_;
+  const Stats::StatName prefix_;
+  bool enabled_;
+  const std::string latency_suffix_ = ".latency";
```

Comment:
if possible make these StatName objects allocated out of your stat_name_set, and use join().

Reply:
Clarification: these should be populated in the construction like:
```
  : total_suffix_(stat_name_set_.add("total"),
    success_suffix_(stat_name_set_.add("suffix"),
    ...
```
You don't need to put these in the maps in StatNameSet, so add() is sufficient.

Reply:
If I do this in construction (ie `total_(stat_name_set_.add("total")),`), does that obviate the need to use `rememberBuiltin()`? Assuming that is the case, is it good enough to have `StatName`s for my prefix, various command names and suffixes, versus having to somehow precache the various grouping beforehand (I didn't see anything for pre-remembering `StatNameVec`s in the code).

So, for example, my constructor could become:

```
RedisCommandStats::RedisCommandStats(Stats::Scope& scope, const std::string& prefix, bool enabled)
    : scope_(scope), stat_name_set_(scope.symbolTable()), 
      prefix_(stat_name_set_.add(prefix)),
      enabled_(enabled), 
      upstream_rq_time_(stat_name_set_.add("upstream_rq_time")),
      latency_(stat_name_set_.add("latency")), ... {}
```

with no need for `createStats(std::string name)`, as the `.add` already takes care of that?

Reply:
Yeah I think so. You don't need a string-to-StatName map for those explicit suffixes as you are keeping the reference to the StatName in your structure, with storage managed by by stat_name_set_.pool_

Reply:
Yeah I think so. You don't need a string-to-StatName map for those explicit suffixes as you are keeping the reference to the StatName in your structure, with storage managed by by stat_name_set_.pool_